CI information in UserAgent header (#14717)
NuGet/Home: accepted/2026/ci-useragent-nuget.md

# User-Agent Telemetry Enrichment for NuGet Commands

> - Author: [@mruizmares](https://github.com/mruizmares)

Reviewer: that's not your account, and URL currently is a 404 😆

> - GitHub Issue: https://github.com/NuGet/Client.Engineering/issues/3467

Reviewer 2: Should we link to private issues in public specs?

Reviewer: IMO no, we shouldn't.

## Summary

> This proposal adds telemetry enrichment to the NuGet User-Agent header for push and restore commands. The enriched User-Agent header includes contextual information about the client environment (GitHub Actions, Azure DevOps) and the current operation (push, restore), enabling better analytics and diagnostics for NuGet server operators.

Reviewer 3: The summary says we are adding the current operation, but that's not anywhere in the proposal.

## Motivation

Currently, NuGet HTTP requests include a basic User-Agent header that identifies the NuGet client version and OS.

By enriching the User-Agent header with CI context information we can a better understanding of CI/CD environment usage and issues.

## Explanation

### Functional explanation

When you run NuGet commands, the HTTP requests now include enriched User-Agent information:

**Before:**
```
NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0)
```

**After (running in GitHub Actions, push command):**
```
NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0) CI/GitHubActions
```

The enrichment happens automatically:
1. **Detection**: NuGet detects if it's running in a known CI/CD environment (GitHub Actions or Azure DevOps) by checking environment variables.

> 3. **User-Agent Enrichment**: HTTP requests include this context in the User-Agent header.

Reviewer 4: ```suggestion
1. **User-Agent Enrichment**: HTTP requests include this context in the User-Agent header.
```

> 3. **User-Agent Enrichment**: HTTP requests include this context in the User-Agent header.

Reviewer 4: Are there any privacy considerations here? I don't see this being PII and it's just basically a switch in the header, so I suspect it's a non-issue?

Author: I suspect it's not an issue since it's not PII, it's system metadata which we are already tracking

### Technical explanation

**CI Detection**: 

> Detects CI/CD environments from environment variables:

Reviewer 2: Consider this:
https://github.com/dotnet/reproducible-builds/blob/8714cf7b775ee8c13933c1e5bfbe1b9d1d9a9be4/src/DotNet.ReproducibleBuilds/DotNet.ReproducibleBuilds.props#L15-L38

Reviewer 2: Ah, Andy has a similar comment below: https://github.com/NuGet/Home/pull/14717/files#r2688273664

| Environment | Variable | Value | ClientId | Info |
|-------------|----------|-------|----------|------|

> | GitHub Actions | `GITHUB_ACTIONS` | "true" | "GitHub" | https://docs.github.com/en/actions/reference/workflows-and-actions/variables?versionId=free-pro-team%40latest&productId=actions |

Reviewer 4: Client machines could also have this set, but I figure we're ok with that likely small margin-of-error?

> | Azure DevOps | `TF_BUILD` | "True" | "AzureDevOps" | https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#system-variables |

Reviewer 3: I know a bunch of CI vendors also set the `CI` env var, should we add that too?

Basically, CI/Undefined or CI/Unknown

#### User-Agent Format

The enriched User-Agent follows the format:
```

> {base-user-agent} NuGet/{ClientId} CI/{CiClient}

Reviewer 2: @agr could you identify if this change will impact NuGet.org statistics processing?

Author: @joelverhagen @agr I updated the format to `{base-user-agent} NuGet/{ClientId; CIClient}`

```

Examples:
- `NuGet xplat/6.10.0 (Microsoft Windows NT) CI/GitHubActions`
- `NuGet xplat/6.10.0 (Linux) CI/AzureDevOps`

## Drawbacks

> 1. **Limited CI/CD Coverage**: Only GitHub Actions and Azure DevOps are detected initially. Other CI systems are not detected.

Reviewer: The SDK already detects CI for telemetry purposes, and supports several other environment variables: https://github.com/dotnet/sdk/blob/1b58d7631b1ed3cd26f7fa7415c075f686ae0f1a/src/Cli/dotnet/Telemetry/CIEnvironmentDetectorForTelemetry.cs#L13

Since NuGet is a component of the SDK, I think it should be consistent with the rest of the SDK.  The technical challenge is how to detect when the SDK changes the list so that NuGet can adapt and use the same values

> ### Why this design?

Reviewer: Why a `CI/GitHubActions` addition, rather than making it `(Linux; CI)`?

Is there prior art from other tools that detect CI?  How do they convey the same information?

It's been years since I've looked into the RFC that defines the user agent string. Is CI/AzureDevOps consistent with the intent for that vs the `(whatever)` part of the string?

Reviewer: Asking a chatbot real quick, it says that `/` is used for product + version, whereas `(...)` is used for description.   CI/GitHubActions doesn't match product+version, so it doesn't feel right to me.  I know that the .NET APIs do some validation on user agent strings, as we've had unhandled exceptions in the past.  Does .NET even allow you to use GitHubActions or AzureDevOps as a version string?

Author: I updated the UserAgent to be `{base-user-agent} NuGet/{ClientId; CIClient}` ex: `NuGet xplat/6.10.0 (Linux; CI: AzureDevOps)`

# User-Agent Telemetry Enrichment for NuGet Commands

- Author: [@mruizmares](https://github.com/martinrrm)
- GitHub Issue: https://github.com/NuGet/Client.Engineering/issues/3467

## Summary

> This proposal adds telemetry enrichment to the NuGet User-Agent header for push and restore commands. The enriched User-Agent header includes contextual information about the client environment (GitHub Actions, Azure DevOps).

Reviewer: In several pull requests on the docs repo and design specs, I've suggested that markdown files should have one sentence per line, as it enriches the pull request experience. I can enrich my feedback by easily commenting on a single line, rather than the full paragraph, and the output when rendered in html is not modified. In the past github didn't highlight changes within long lines, so splitting markdown paragraphs to short lines enriches the diff experience, but that's more helpful in the docs repo where modifications are common.

also nitpick: I have no problem with the word enrich, but it's not a very commonly used word, so using it twice in two sentences feels weird 😁

## Motivation

Currently, NuGet HTTP requests include a basic User-Agent header that identifies the NuGet client version and OS.

By enriching the User-Agent header with CI context information we can a better understanding of CI/CD environment usage and issues.

## Explanation

### Functional explanation

When you run NuGet commands, the HTTP requests now include enriched User-Agent information:

**Before:**
```
NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0)
```

**After (running in GitHub Actions, push command):**
```

> NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0; CI: GithubActions)

Reviewer 2: ```suggestion
NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0; CI: GitHub Actions)
```

```

The enrichment happens automatically:
1. **Detection**: NuGet detects if it's running in a known CI/CD environment (GitHub Actions or Azure DevOps) by checking environment variables.

> 2. **User-Agent Enrichment**: HTTP requests include this context in the User-Agent header.

Reviewer 3: Will there be a header if no CI was detected?
Is that gonna be the assumption by omission or are we gonna be explicit about it?

Author: If no CI is detected, we are going to omit adding the "CI" to the header

### Technical explanation

**CI Detection**: 

Detects CI/CD environments from environment variables:

| Environment | Variable | Value | ClientId | Info |
|-------------|----------|-------|----------|------|

> | GitHub Actions | `GITHUB_ACTIONS` | "true" | "GitHub" | https://docs.github.com/en/actions/reference/workflows-and-actions/variables?versionId=free-pro-team%40latest&productId=actions |

Reviewer 2: ```suggestion
| GitHub Actions | `GITHUB_ACTIONS` | "true" | "GitHub Actions" | https://docs.github.com/en/actions/reference/workflows-and-actions/variables?versionId=free-pro-team%40latest&productId=actions |
```

> | Azure DevOps | `TF_BUILD` | "True" | "AzureDevOps" | https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#system-variables |

Reviewer 2: ```suggestion
| Azure DevOps | `TF_BUILD` | "True" | "Azure DevOps" | https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#system-variables |
```

| AppVeyor | `APPVEYOR` | "True" | "AppVeyor" | https://www.appveyor.com/docs/environment-variables/ |
| Travis | `TRAVIS` | "True" | "Travis CI" | https://docs.travis-ci.com/user/environment-variables/#default-environment-variables |
| CircleCI | `CIRCLECI` | "True" | "CircleCI" | https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables |
| AWS CodeBuild | `CODEBUILD_BUILD_ID` | != '' | AWS CodeBiuld | https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-env-vars.html |
| Jenkins | `BUILD_ID` AND `BIULD_URL` | != '' AND != '' | Jenkins | https://www.jenkins.io/doc/book/pipeline/jenkinsfile/#using-environment-variable |
| Google Cloud | `BUILD_ID` AND `PROJECT_ID` | != '' AND != '' | Google Cloud | https://cloud.google.com/build/docs/configuring-builds/substitute-variable-values |
| TeamCity | `TEAMCITY_VERSION` | != '' | TeamCity | https://www.jetbrains.com/help/teamcity/predefined-build-parameters.html#Server+Build+Properties |

> | JetBrains | `JB_SPACE_API_URL` | != '' | JetBrains | https://www.jetbrains.com/help/space/automation-environment-variables.html#general |

Reviewer 2: ```suggestion
| JetBrains | `JB_SPACE_API_URL` | != '' | JetBrains Space | https://www.jetbrains.com/help/space/automation-environment-variables.html#general |
```

| CI | `CI` | "True" | "CI" | A general-use flag | |

#### User-Agent Format

The enriched User-Agent follows the format:
```

> {base-user-agent} NuGet/{ClientId; CI: ([^\);]+)}

Reviewer: FWIW, I find this very difficult to understand. I know only the very basics of regex. The examples below help, but I personally don't find this line helpful in understanding the proposed change. I suspect that it could be explained in a more clear way.

> We'll add `CI: {detected environment}` to the user agent, following the operating system information

Reviewer 2: I don't think we should have the regex in the spec. I provided it in the comment above as an example. I agree with you.

```

Examples:

> - `NuGet xplat/6.10.0 (Microsoft Windows NT; GitHubActions)`

Reviewer 2: IMHO we should distinguish the CI values with a "CI" tag, that ensures the set of CI strings will not overlap with other strings put in this parenthesis (called a comment in UA language).

So, maybe `NuGet xplat/6.10.0 (Microsoft Windows NT; CI: GitHub Actions)`

Reviewer 2: You could imagine this being used via a regex like `CI: ([^\);]+)`

- `NuGet xplat/6.10.0 (Linux; CI: AzureDevOps)`

> - `NuGet xplat/6.10.0 (Linux; CI: CI)`

Reviewer 2: ```suggestion
- `NuGet xplat/6.10.0 (Linux; CI)`
```
might be a little less ugly :)

Reviewer 2: or `CI: other`

# User-Agent Telemetry Enrichment for NuGet Commands

- Author: [@mruizmares](https://github.com/martinrrm)
- GitHub Issue: https://github.com/NuGet/Home/issues/14740

## Summary

This proposal adds telemetry to enrich the NuGet User-Agent header for push and restore commands. 

> The User-Agent header includes contextual information about the client environment (GitHub Actions, Azure  DevOps).

Reviewer: ```suggestion
This proposal adds CI environment to the User-Agent header for push and restore commands. 
```

For me, "adds telemetry to enrich the User-Agent header" just feels weird. I think simpler language more clearly sets expectations and is easier to read.

## Motivation

Currently, NuGet HTTP requests include a basic User-Agent header that identifies the NuGet client version and OS.

> By enriching the User-Agent header with CI context information we can a better understanding of CI/CD environment usage and issues.

Reviewer: ```suggestion
By enriching the User-Agent header with CI context information we can get a better understanding of CI/CD environment usage and issues.
```

## Explanation

### Functional explanation

When you run NuGet commands, the HTTP requests now include enriched User-Agent information:

**Before:**
```
NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0)
```

**After (running in GitHub Actions, push command):**
```
NuGet Command Line/6.10.0 (Microsoft Windows NT 10.0.22631.0; CI: GithubActions)
```

The enrichment happens automatically:
1. **Detection**: NuGet detects if it's running in a known CI/CD environment (GitHub Actions or Azure DevOps) by checking environment variables.

> If no CI is detected, the User-Agent header won't have a `CI` information.

Reviewer: ```suggestion
If no CI is detected, the User-Agent header won't have `CI` information.
```

| CI | `CI` | "True" | "other" | A general-use flag | |

#### User-Agent Format

The enriched User-Agent follows the format:
```
{base-user-agent} NuGet/{ClientId; CI: {detected environment}}
```

Examples:
- `NuGet xplat/6.10.0 (Microsoft Windows NT; CI: GitHubActions)`
- `NuGet xplat/6.10.0 (Linux; CI: AzureDevOps)`
- `NuGet xplat/6.10.0 (Linux; CI: other)`

## Drawbacks

1. **Limited CI/CD Coverage**: We have to manually update the table if we want to track a new specific CI 
environment. 
We are considering the exising CI environments that [dotnet sdk is tracking](https://github.com/dotnet/sdk/blob/1b58d7631b1ed3cd26f7fa7415c075f686ae0f1a/src/Cli/dotnet/Telemetry/CIEnvironmentDetectorForTelemetry.cs#L13).
Other CI environment will be considered as "CI".

### Why this design?

1. **User-Agent vs Custom Headers**: Enriching the existing User-Agent header is preferred over custom headers because:
   - It works with all HTTP infrastructure without configuration
   - It's logged by most web servers by default
   - It follows established conventions for client identification

### Alternatives considered

## Unresolved Questions

> 1. **User Opt-out**: Should there be a way for users to disable the enriched User-Agent? What would the mechanism be (environment variable, config setting)?

Reviewer: Is the intention to resolve this question before the spec is merged as approved, or leave it unresolved?
